source/val/ValidationState.cpp

```diff
@@ -357,4 +365,91 @@ spv_result_t ValidationState_t::RegisterFunctionEnd() {
   return SPV_SUCCESS;
 }
 
+ValidationState_t::UseDefTracker::UseDefTracker(ValidationState_t* state)
+    : module_(state) {}
+
+void ValidationState_t::UseDefTracker::AddDef(const spv_id_info_t& def) {
+  all_defs_[def.id] = def;
+  if (module_->in_block()) {
+    auto block = module_->get_current_function()->get_current_block();
+    block_defs_[block->get_id()].push_back(def.id);
+  } else {
+    module_defs_.insert(def.id);
+  }
+}
+
+void ValidationState_t::UseDefTracker::AddUse(uint32_t id, bool is_phi) {
+  all_uses_.insert(id);
+  if (module_->in_block()) {
+    auto block = module_->get_current_function()->get_current_block();
+    block_uses_[block->get_id()].insert({id, is_phi});
+  }
+}
+
+std::pair<bool, spv_id_info_t> ValidationState_t::UseDefTracker::FindDef(
+    uint32_t id) const {
+  if (all_defs_.count(id) == 0) {
+    return std::make_pair(false, spv_id_info_t{});
+  } else {
+    /// We are in a const function, so we cannot use defs.operator[]().
+    /// Luckily we know the key exists, so defs_.at() won't throw an
+    /// exception.
+    return std::make_pair(true, all_defs_.at(id));
+  }
+}
+
+std::unordered_set<uint32_t>
+ValidationState_t::UseDefTracker::FindUsesWithoutDefs() const {
+  auto diff = all_uses_;
+  for (const auto d : all_defs_) diff.erase(d.first);
+  return diff;
+}
+
+void ValidationState_t::UseDefTracker::UpdateDefs(Function* func) {
+  auto& ordered_blocks = func->get_blocks();
+
+  // Because dominators must appear before the blocks they dominate in the
+  // binary , we only need to visit the immidiate dominators to propagate the
+  // ids throughout the CFG
+  for (auto block : ordered_blocks) {
+    auto& block_defs = block_defs_[block->get_id()];
+    auto idom = block->GetImmediateDominator();
+    if (block->is_reachable() && block != idom) {
```

Comment:
I'm a little fuzzy on whether this idom variable is always non-null.  I think it is, once I've traced back through the various algorithms.  Could do with an assert here.

Normally the idom of a node strictly dominates the node, which means it can't be the node.
For example, is the idom of the entry block the entry block itself?
The blurb on CalculateDominators is too vague to tell this apart.